### ISSUE TITLE
feature request: kwargs for getindex

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#25631]
Current behavior is to treat `=` as assignment inside `[]`
```julia
julia> Meta.@lower x[a=1] = 3
:(begin
      a = 1
      setindex!(x, 3, 1)
      return 3
  end)
```

But for consistency with `()`, I would like to have it parse as a kwcall
```julia
setindex!(x, 3; a=1)
```

Currently, that has to be written:
```julia
x[(a = 1,)] = 3
```
which is much less nice.

**Comment by stevengj** [comment]
I've used things like `a[i += 1]` to simultaneously increment a counter and do `getindex` or `setindex`, but presumably this would continue to work, since keywords happen in parsing and  `+=` is dealt with in lowering.

**Comment by StefanKarpinski** [comment]
Agree, this seems like a really bizarre misfeature request to me. The fact that indexing is lowered to a function call is incidental, indexing syntax is not function call syntax.

**Comment by roshanshariff** [comment]
I wanted to use this syntax for multidimensional arrays with "named" dimensions (as in the Python library [xarray](http://xarray.pydata.org/)). So if `M` is a two-dimensional array with dimensions named `x` and `y`, I'd like to be able to write `M[x=3, y=2]` to index the (3, 2) element. Currently there's no way to use the subscript notation for `getindex(A; dims...)`.

**Comment by JeffBezanson** [comment]
That does seem more useful to me than doing an assignment inside an indexing expression (`a[b+=1]` is indeed a separate case and would continue to work).

**Comment by Keno** [comment]
I find that argument convincing.

**Comment by martinholters** [comment]
In function calls, we also allow e.g. `x+=1` but---obviously---treat `x=1` as a keyword argument, so doing the same for indexing seems fine.

**Comment by StefanKarpinski** [comment]
Is this on the wrong milestone?

**Comment by mbauman** [comment]
Re-opening to track the possibility of adding this feature in 1.x now that the old behavior has been deprecated.

**Comment by c42f** [comment]
It appears this feature was implemented in 1.0 as part of 28366:

```julia
julia> dump(:(x[a=1]))
Expr
  head: Symbol ref
  args: Array{Any}((2,))
    1: Symbol x
    2: Expr
      head: Symbol kw
      args: Array{Any}((2,))
        1: Symbol a
        2: Int64 1
```
